src/diffusers/image_processor.py

```diff
@@ -0,0 +1,173 @@
+        """
+        Convert a numpy image to a pytorch tensor
+        """
+        images = images.cpu().permute(0, 2, 3, 1).float().numpy()
+        return images
+
+    @staticmethod
+    def normalize(images):
+        """
+        Normalize an image array to [-1,1]
+        """
+        return 2.0 * images - 1.0
+
+    def resize(self, images: PIL.Image.Image) -> PIL.Image.Image:
+        """
+        Resize an PIL image. Both height and width will be resized to integer multiple of vae_scale_factor
+        """
+        w, h = images.size
+        w, h = map(lambda x: x - x % self.vae_scale_factor, (w, h))  # resize to integer multiple of vae_scale_factor
+        images = images.resize((w, h), resample=PIL_INTERPOLATION[self.resample])
+        return images
+
+    def encode(
+        self,
+        image: Union[torch.FloatTensor, PIL.Image.Image, np.ndarray],
+    ) -> torch.Tensor:
+        """
+        Preprocess the image input, accpet formats in PIL images, numpy arrays or pytorch tensors"
+        """
+        supported_formats = (PIL.Image.Image, np.ndarray, torch.Tensor)
+        if isinstance(image, supported_formats):
+            image = [image]
+        elif isinstance(image, list) and all(isinstance(i, supported_formats) for i in image):
+            image = image
+        else:
+            raise ValueError(
+                "incorrect image format is used - currently we only support PIL image, numpy array or pytorch tensor"
+            )
+
+        if isinstance(image[0], PIL.Image.Image):
+            if self.do_resize:
+                image = [self.resize(i) for i in image]
+            image = [np.array(i).astype(np.float32) / 255.0 for i in image]
+
+        if isinstance(image[0], np.ndarray):
+            image = self.numpy_to_pt(np.stack(image, axis=0))
+        elif not isinstance(image, torch.Tensor) and isinstance(image[0], torch.Tensor):
+            image = torch.cat(image, dim=0)
+
+        if image.ndim == 5:
+            image = image.squeeze(0)
+        _, _, height, width = image.shape
+        if self.do_resize and (height % self.vae_scale_factor != 0 or width % self.vae_scale_factor != 0):
+            raise ValueError(
+                f"the height and width of image have to be divisible by {self.vae_scale_factor} but are {height} and {width}."
+            )
+
+        # expected range [0,1], normalize to [-1,1]
+        do_normalize = self.do_normalize
+        if image.min() < 0:
+            warnings.warn(
+                "Passing `image` as torch tensor with value range in [-1,1] is deprecated. The expected value range for image tensor is [0,1] "
+                f"when passing as pytorch tensor or numpy Array. You passed `image` with value range [{image.min()},{image.max()}]",
+                FutureWarning,
+            )
+            do_normalize = False
+
+        if do_normalize:
+            image = self.normalize(image)
+
+        return image
+
+    def decode(
```

Comment:
Shouldn't we denormalize here (if appropriate), to return the range to `[0, 1]`?

Reply:
currently, it's denormalized inside `decode_latent` - I think we can move it to the image processor, but I'm not sure how the decoding part of the image processor fits in the pipeline - I tried to refactor the img2img pipeline with it, but it seems that we can't abstract the post-processing away from the pipeline if we don't move the safty_checker to image processor